File: apps/api/app/models/due_diligence.py
"""Due Diligence Checklist models."""

from __future__ import annotations
import uuid
from datetime import datetime
from typing import Any

from sqlalchemy import Boolean, Float, ForeignKey, Index, Integer, String, Text, UniqueConstraint
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.models.base import BaseModel


class DDChecklistTemplate(BaseModel):
    __tablename__ = "dd_checklist_templates"
    __table_args__ = (
        Index("ix_dd_template_type_stage", "asset_type", "deal_stage"),
    )

    asset_type: Mapped[str] = mapped_column(String(50), nullable=False, index=True)
    deal_stage: Mapped[str] = mapped_column(String(50), nullable=False, index=True)
    jurisdiction_group: Mapped[str | None] = mapped_column(String(20), nullable=True)
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    version: Mapped[int] = mapped_column(Integer, default=1, server_default="1")
    is_active: Mapped[bool] = mapped_column(Boolean, default=True, server_default="true")
    created_by: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id", ondelete="SET NULL"), nullable=True
    )


class DDChecklistItem(BaseModel):
    __tablename__ = "dd_checklist_items"

    template_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("dd_checklist_templates.id", ondelete="CASCADE"),
        nullable=False, index=True
    )
    category: Mapped[str] = mapped_column(String(50), nullable=False)
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    requirement_type: Mapped[str] = mapped_column(String(30), nullable=False)
    required_document_types: Mapped[list | None] = mapped_column(JSONB, nullable=True)
    verification_criteria: Mapped[str | None] = mapped_column(Text, nullable=True)
    priority: Mapped[str] = mapped_column(String(20), default="required", server_default="required")
    sort_order: Mapped[int] = mapped_column(Integer, default=0, server_default="0")
    estimated_time_hours: Mapped[float | None] = mapped_column(Float, nullable=True)
    regulatory_reference: Mapped[str | None] = mapped_column(String(255), nullable=True)


class DDProjectChecklist(BaseModel):
    __tablename__ = "dd_project_checklists"

    project_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("projects.id", ondelete="CASCADE"),
        nullable=False, index=True
    )
    org_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("organizations.id", ondelete="CASCADE"),
        nullable=False
    )
    template_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("dd_checklist_templates.id"),
        nullable=False
    )
    investor_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id", ondelete="SET NULL"), nullable=True
    )
    status: Mapped[str] = mapped_column(String(20), default="in_progress", server_default="in_progress")
    completion_percentage: Mapped[float] = mapped_column(Float, default=0.0, server_default="0")
    total_items: Mapped[int] = mapped_column(Integer, default=0, server_default="0")
    completed_items: Mapped[int] = mapped_column(Integer, default=0, server_default="0")
    custom_items: Mapped[list] = mapped_column(JSONB, default=list, server_default="[]")


class DDItemStatus(BaseModel):
    __tablename__ = "dd_item_statuses"
    __table_args__ = (
        UniqueConstraint("checklist_id", "item_id", name="uq_dd_item_status"),
        Index("ix_dd_item_status_checklist", "checklist_id"),
    )

    checklist_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("dd_project_checklists.id", ondelete="CASCADE"),
        nullable=False
    )
    item_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("dd_checklist_items.id", ondelete="CASCADE"),
        nullable=False
    )
    status: Mapped[str] = mapped_column(String(20), default="pending", server_default="pending")
    satisfied_by_document_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("documents.id", ondelete="SET NULL"), nullable=True
    )
    ai_review_result: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
    reviewer_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id", ondelete="SET NULL"), nullable=True
    )
    reviewer_notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    reviewed_at: Mapped[datetime | None] = mapped_column(nullable=True)

File: apps/web/src/lib/due-diligence.ts
/**
 * Due Diligence Checklist — types and React Query hooks.
 *
 * Types mirror the FastAPI Pydantic schemas. Hooks wrap axios calls
 * with React Query for caching, optimistic updates, and state management.
 */

import { useQuery, useMutation, useQueryClient } from "@tanstack/react-query";
import { api } from "@/lib/api";

// ── Types ────────────────────────────────────────────────────────────────────

export type DDItemStatus =
  | "pending"
  | "in_review"
  | "satisfied"
  | "partially_met"
  | "not_met"
  | "waived";

export type DDItemPriority = "required" | "recommended" | "optional";

export interface DDTemplateResponse {
  id: string;
  asset_type: string;
  deal_stage: string;
  jurisdiction_group: string | null;
  name: string;
  description: string | null;
  version: number;
  item_count: number;
}

export interface DDItemResponse {
  id: string;
  template_id: string;
  category: string;
  name: string;
  description: string | null;
  requirement_type: string;
  required_document_types: string[] | null;
  verification_criteria: string | null;
  priority: DDItemPriority;
  sort_order: number;
  estimated_time_hours: number | null;
  regulatory_reference: string | null;
}

export interface DDChecklistItemFull {
  // Item fields
  item_id: string;
  template_id: string;
  category: string;
  name: string;
  description: string | null;
  requirement_type: string;
  required_document_types: string[] | null;
  verification_criteria: string | null;
  priority: DDItemPriority;
  sort_order: number;
  estimated_time_hours: number | null;
  regulatory_reference: string | null;
  // Status fields
  status_id: string | null;
  status: DDItemStatus;
  satisfied_by_document_id: string | null;
  ai_review_result: {
    satisfied?: boolean;
    confidence?: number;
    summary?: string;
    gaps?: string[];
    recommendation?: string;
    error?: string;
  } | null;
  reviewer_notes: string | null;
  reviewed_at: string | null;
}

export interface DDChecklistResponse {
  id: string;
  project_id: string;
  org_id: string;
  template_id: string;
  investor_id: string | null;
  status: string;
  completion_percentage: number;
  total_items: number;
  completed_items: number;
  custom_items: CustomItem[];
  items_by_category: Record<string, DDChecklistItemFull[]>;
  created_at: string;
  updated_at: string;
}

export interface CustomItem {
  id: string;
  name: string;
  category: string;
  description: string | null;
  priority: DDItemPriority;
  status: DDItemStatus;
  created_at: string;
}

// ── Request types ─────────────────────────────────────────────────────────────

export interface GenerateChecklistRequest {
  project_id: string;
  investor_id?: string;
}

export interface UpdateItemStatusRequest {
  status: DDItemStatus;
  notes?: string;
  document_id?: string;
}

export interface AddCustomItemRequest {
  name: string;
  category: string;
  description?: string;
  priority?: DDItemPriority;
}

export interface TriggerAIReviewRequest {
  document_id: string;
}

// ── Query keys ────────────────────────────────────────────────────────────────

export const DD_KEYS = {
  templates: (assetType?: string) =>
    ["dd-templates", assetType].filter(Boolean) as string[],
  template: (id: string) => ["dd-template", id],
  checklists: (projectId?: string) =>
    ["dd-checklists", projectId].filter(Boolean) as string[],
  checklist: (id: string) => ["dd-checklist", id],
};

// ── Hooks ─────────────────────────────────────────────────────────────────────

/**
 * List templates, optionally filtered by asset_type.
 */
export function useTemplates(assetType?: string) {
  return useQuery({
    queryKey: DD_KEYS.templates(assetType),
    queryFn: async () => {
      const params = assetType ? `?asset_type=${assetType}` : "";
      const res = await api.get<DDTemplateResponse[]>(
        `/due-diligence/templates${params}`
      );
      return res.data;
    },
  });
}

/**
 * Get a single template with items.
 */
export function useTemplate(templateId: string | undefined) {
  return useQuery({
    queryKey: DD_KEYS.template(templateId ?? ""),
    queryFn: async () => {
      const res = await api.get(`/due-diligence/templates/${templateId}`);
      return res.data;
    },
    enabled: !!templateId,
  });
}

/**
 * List checklists for a project.
 */
export function useProjectChecklist(projectId: string | undefined) {
  return useQuery({
    queryKey: DD_KEYS.checklists(projectId),
    queryFn: async () => {
      const res = await api.get<DDChecklistResponse[]>(
        `/due-diligence/checklists?project_id=${projectId}`
      );
      // Return first checklist if it exists
      return res.data[0] ?? null;
    },
    enabled: !!projectId,
  });
}

/**
 * Get a specific checklist by ID.
 */
export function useChecklist(checklistId: string | undefined) {
  return useQuery({
    queryKey: DD_KEYS.checklist(checklistId ?? ""),
    queryFn: async () => {
      const res = await api.get<DDChecklistResponse>(
        `/due-diligence/checklists/${checklistId}`
      );
      return res.data;
    },
    enabled: !!checklistId,
  });
}

/**
 * Generate a checklist for a project.
 */
export function useGenerateChecklist() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: async (data: GenerateChecklistRequest) => {
      const res = await api.post<DDChecklistResponse>(
        "/due-diligence/checklists/generate",
        data
      );
      return res.data;
    },
    onSuccess: (_, variables) => {
      queryClient.invalidateQueries({
        queryKey: DD_KEYS.checklists(variables.project_id),
      });
    },
  });
}

/**
 * Update the status of a checklist item.
 */
export function useUpdateItemStatus(checklistId: string) {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: async ({
      itemId,
      ...data
    }: UpdateItemStatusRequest & { itemId: string }) => {
      const res = await api.put(
        `/due-diligence/checklists/${checklistId}/items/${itemId}/status`,
        data
      );
      return res.data;
    },
    onSuccess: () => {
      queryClient.invalidateQueries({
        queryKey: DD_KEYS.checklist(checklistId),
      });
    },
  });
}

/**
 * Add a custom item to a checklist.
 */
export function useAddCustomItem(checklistId: string) {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: async (data: AddCustomItemRequest) => {
      const res = await api.post(
        `/due-diligence/checklists/${checklistId}/items/add`,
        data
      );
      return res.data;
    },
    onSuccess: () => {
      queryClient.invalidateQueries({
        queryKey: DD_KEYS.checklist(checklistId),
      });
    },
  });
}

/**
 * Trigger AI review for a document against a checklist item.
 */
export function useTriggerAIReview(checklistId: string) {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: async ({
      itemId,
      document_id,
    }: {
      itemId: string;
      document_id: string;
    }) => {
      const res = await api.post(
        `/due-diligence/checklists/${checklistId}/items/${itemId}/review`,
        { document_id }
      );
      return res.data;
    },
    onSuccess: () => {
      queryClient.invalidateQueries({
        queryKey: DD_KEYS.checklist(checklistId),
      });
    },
  });
}
